src/ripple/rpc/impl/ServerHandlerImp.cpp

```diff
@@ -545,164 +557,243 @@ ServerHandlerImp::processRequest (Port const& port,
-    // Determine role/usage so we can charge for invalid requests
-    Json::Value const& method = jsonRPC [jss::method];
-
-    auto role = Role::FORBID;
-    auto required = RPC::roleRequired(method.asString());
-    if (jsonRPC.isMember(jss::params) &&
-        jsonRPC[jss::params].isArray() &&
-        jsonRPC[jss::params].size() > 0 &&
-        jsonRPC[jss::params][Json::UInt(0)].isObject())
-    {
-        role = requestRole(required, port, jsonRPC[jss::params][Json::UInt(0)],
-            remoteIPAddress, user);
-    }
-    else
+    bool batch = false;
+    if (jsonOrig.isMember(jss::method) && jsonOrig[jss::method] == "batch")
+        batch = true;
+    auto size = batch ? jsonOrig[jss::params].size() : 1;
+    Json::Value reply(batch ? Json::arrayValue : Json::objectValue);
+    auto const start (std::chrono::high_resolution_clock::now ());
+    for (unsigned i = 0; i < size; ++i)
     {
-        role = requestRole(required, port, Json::objectValue,
-            remoteIPAddress, user);
-    }
+        Json::Value const& jsonRPC = batch ? jsonOrig[jss::params][i] : jsonOrig;
+        /* ---------------------------------------------------------------------- */
+        // Determine role/usage so we can charge for invalid requests
+        Json::Value const& method = jsonRPC [jss::method];
+
+        auto role = Role::FORBID;
+        auto required = RPC::roleRequired(method.asString());
+        if (jsonRPC.isMember(jss::params) &&
+            jsonRPC[jss::params].isArray() &&
+            jsonRPC[jss::params].size() > 0 &&
+            jsonRPC[jss::params][Json::UInt(0)].isObject())
+        {
+            role = requestRole(required, port, jsonRPC[jss::params][Json::UInt(0)],
+                remoteIPAddress, user);
+        }
+        else
+        {
+            role = requestRole(required, port, Json::objectValue,
+                remoteIPAddress, user);
+        }
 
-    Resource::Consumer usage;
-    if (isUnlimited(role))
-    {
-        usage = m_resourceManager.newUnlimitedEndpoint(
-            remoteIPAddress.to_string());
-    }
-    else
-    {
-        usage = m_resourceManager.newInboundEndpoint(remoteIPAddress);
-        if (usage.disconnect())
+        Resource::Consumer usage;
+        if (isUnlimited(role))
         {
-            HTTPReply(503, "Server is overloaded", output, rpcJ);
-            return;
+            usage = m_resourceManager.newUnlimitedEndpoint(
+                remoteIPAddress.to_string());
+        }
+        else
+        {
+            usage = m_resourceManager.newInboundEndpoint(remoteIPAddress);
+            if (usage.disconnect())
+            {
+                if (!batch)
+                {
+                    HTTPReply(503, "Server is overloaded", output, rpcJ);
+                    return;
+                }
+                Json::Value r = jsonRPC;
+                r[jss::error] = make_json_error(-32604, "Server is overloaded");
```

Comment:
I had to look up these numbers to figure out where they came from. Might be helpful to define some constexpr data to use here, but I'm not hung up on it. Alternatively a link to the spec somewhere here could clue us in as to the source of the values.

Reply:
Good suggestion, thanks.